### PR TITLE
Implement multibase for C++

### LINKED ISSUE #168: Implement base64 support in C++
Currently we use b58, however due to `O(n^2)` nature of b58 coding, this isn't ideal for large strings.

Let's use base64 instead.

We can't change the default in C++ from base58 to base64 until Golang supports base64.  So we should change the C++ and Golang defaults together as part of #169.

### LINKED ISSUE #169: Implement base64 support in Golang, set variable_blob to use base64
This ticket is to use the libp2p multibase library for the Golang side of #168.

We can't change the default in C++ to base64 until Golang supports it.  So we should change the defaults for both C++ and Golang to base64 as part of this ticket.

### PR DESCRIPTION
**Pull request by theoreticalbts**
Closes #168.
Resolves #169.

### CODE REVIEW

**Review by sgerbino — CHANGES_REQUESTED** [review]
- We need more extensive coverage via testing of new code
- We shouldn't remove the bitcoin copyright
- The tests should be implemened as a Boost Test and run in CI

**Comment by theoreticalbts** [comment]
The copyright wasn't removed, it was moved lower in the file.

For more clarity about what copyright applies to what code, I'm going to update the branch and put the code from Bitcoin in a separate header file from our code.

**Comment by theoreticalbts** [comment]
Updated this PR:

- Addressed copyright issue
- Added tests
- Fixed several bugs uncovered by the tests

Also added a Monte Carlo test which you can run forever by:

```
    ../programs/test_multibase/multibase_mc.py | programs/test_multibase/test_multibase
```

You can insert a filter, e.g. `-head 1000` to do 1000 cases, or `pv -l` to see a progress bar.  The MC test has passed over 5 million cases with no failures.

File: programs/koinos-types/lang/koinos_codegen_cpp/rt/json.hpp [hunk 1]
```diff
@@ -210,30 +210,19 @@ inline void from_json( const json& j, vector< T >& v, uint32_t depth )
 // variable length blob
 inline void to_json( json& j, const variable_blob& v )
 {
-   string base58;
-   util::encode_base58( base58, v );
-   j = std::move( 'z' + base58 );
+   std::vector<char> dest;
+   util::encode_multibase( v.data(), v.size(), dest );
```
Comment (mvandeberg) [inline programs/koinos-types/lang/koinos_codegen_cpp/rt/json.hpp:214]: We need to make the default variable_blob encoding base64

File: programs/koinos-types/lang/koinos_codegen_cpp/rt/json.hpp [hunk 2]
```diff
@@ -304,29 +293,20 @@ inline void from_json( const json& j, array< T, N >& v, uint32_t depth )
 template< size_t N >
 inline void to_json( json& j, const fixed_blob< N >& v )
 {
-   string base58;
-   util::encode_base58( base58, v );
-   j = std::move( 'z' + base58 );
+   std::vector<char> dest;
+   util::encode_multibase( v.data(), v.size(), dest );
```
Comment (mvandeberg) [inline programs/koinos-types/lang/koinos_codegen_cpp/rt/json.hpp:297]: We need to make the default fixed_blob encoding base64

File: programs/koinos-types/lang/koinos_codegen_cpp/rt/multihash_json.hpp [hunk 3]
```diff
@@ -16,23 +16,21 @@ inline void to_json( json& j, const multihash& v )
...
    std::stringstream ss;
    to_binary( ss, v );
-   std::string serialized_b58;
-   util::encode_base58( serialized_b58, ss.str() );
-   j = std::string("z") + serialized_b58;
+   // Could we avoid this copy?
+   std::string serialized_bytes = ss.str();
+
+   std::vector<char> encoded;
+   util::encode_multibase( serialized_bytes.c_str(), serialized_bytes.size(), encoded );
```
Comment (mvandeberg) [inline programs/koinos-types/lang/koinos_codegen_cpp/rt/multihash_json.hpp:23]: I know the default encoding is base58, but should we make that explicit here?

File: programs/koinos-types/lang/koinos_codegen_cpp/rt/util/multibase.hpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      default:
+         throw base_decode_error( "Unsupported multibase specifier" );
+   }
+   // Is there an easier way to convert char to a string?
+   dest.resize(temp.size()+1);
+   dest[0] = base;
+   memcpy( dest.data()+1, temp.data(), temp.size() );
+}
+
+/*
```
Comment (mvandeberg) [inline programs/koinos-types/lang/koinos_codegen_cpp/rt/util/multibase.hpp:77]: Is this commented code needed anymore? If not, can we clean it up?

File: tests/cpp/src/multibase_test.cpp [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+   std::cerr << std::endl;
+}
+
+void dump_str( const std::vector< char >& s )
+{
+   std::cerr << "\"" << std::string( s.begin(), s.end() ) << "\"" << std::endl;
+}
+
+
+BOOST_AUTO_TEST_CASE( multibase_test )
```
Comment (mvandeberg) [inline tests/cpp/src/multibase_test.cpp:42]: These are all positive test cases. We need some specific negative test cases.

Some examples:

Illegal characters in an encoding.
Incorrect lengths (base64 with/without padding)

**Review by mvandeberg — APPROVED** [review]

**Pull request merged**